File: backend/api/routes/datasets.py
# ============================================================
# Dataset API routes
# POST /api/datasets/generate  — generate scenario batch
# GET  /api/datasets/list      — list existing scenarios
# GET  /api/datasets/{id}      — fetch single scenario JSON
# ============================================================

from __future__ import annotations

import logging
import pathlib

from fastapi import APIRouter, HTTPException, Query

logger = logging.getLogger(__name__)

from backend.api.schemas import (
    GenerateRequest,
    GenerateResponse,
    ScenarioData,
    ScenarioListItem,
    ScenarioListResponse,
)
from backend.datasets.generator import (
    generate_batch,
    list_scenarios,
    load_scenario,
)

router = APIRouter()

_DEFAULT_DATA_DIR = pathlib.Path(__file__).parent.parent.parent.parent / "data" / "scenarios"


@router.post("/generate", response_model=GenerateResponse)
async def generate_scenarios(request: GenerateRequest) -> GenerateResponse:
    """
    Generate a batch of deterministic scenario JSON files.
    Each scenario is stored under data/scenarios/<split>/<scenario_id>.json.
    """
    try:
        result = generate_batch(request)
        return result
    except Exception as exc:
        logger.exception("Scenario generation failed")
        raise HTTPException(status_code=500, detail="Scenario generation failed. Check server logs.") from exc


@router.get("/list", response_model=ScenarioListResponse)
async def list_scenario_files(
    split: str | None = Query(None, description="Filter by split: train, val, test"),
) -> ScenarioListResponse:
    """Return metadata for all existing scenario files."""
    paths = list_scenarios(split=split)
    items: list[ScenarioListItem] = []

    for p in paths:
        try:
            scenario = load_scenario(p)
            items.append(ScenarioListItem(
                scenario_id=scenario.meta.scenario_id,
                split=scenario.meta.split,
                difficulty=scenario.meta.difficulty,
                seed=scenario.meta.seed,
                bus_count=scenario.meta.bus_count,
                episode_length=scenario.meta.episode_length,
                event_count=len(scenario.passenger_events),
            ))
        except Exception:
            # Skip corrupt files — log but don't crash
            continue

    return ScenarioListResponse(scenarios=items, total=len(items))


@router.get("/{scenario_id}", response_model=ScenarioData)
async def get_scenario(scenario_id: str) -> ScenarioData:
    """Fetch a single scenario by ID (searches all splits)."""
    paths = list_scenarios()
    for p in paths:
        if p.stem == scenario_id:
            try:
                return load_scenario(p)
            except Exception as exc:
                logger.exception("Failed to load scenario '%s'", scenario_id)
                raise HTTPException(status_code=500, detail="Failed to load scenario. Check server logs.") from exc

    raise HTTPException(status_code=404, detail=f"Scenario '{scenario_id}' not found")

File: src/api/datasets.ts
// ============================================================
// Datasets API — /api/datasets/*
// ============================================================

import { apiFetch } from './client';

export interface GenerateRequest {
  count?: number;
  seed_start?: number;
  split?: 'train' | 'val' | 'test';
  difficulty?: 'easy' | 'medium' | 'hard' | 'stress';
  bus_count?: number;
  episode_length?: number;
  spawn_rate?: number;
  passenger_capacity?: number;
  reboard_enabled?: boolean;
  charging_enabled?: boolean;
  low_battery_threshold?: number;
  charge_duration?: number;
  battery_drain_rate?: number;
}

export interface GenerateResponse {
  generated: number;
  scenario_ids: string[];
  output_dir: string;
}

export interface ScenarioListItem {
  scenario_id: string;
  split: string;
  difficulty: string;
  seed: number;
  bus_count: number;
  episode_length: number;
  event_count: number;
}

export interface ScenarioListResponse {
  scenarios: ScenarioListItem[];
  total: number;
}

export function generateScenarios(req: GenerateRequest = {}): Promise<GenerateResponse> {
  return apiFetch('/datasets/generate', {
    method: 'POST',
    body: JSON.stringify(req),
  });
}

export function listScenarios(split?: string): Promise<ScenarioListResponse> {
  const qs = split ? `?split=${split}` : '';
  return apiFetch(`/datasets/list${qs}`);
}
